src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp

```diff
@@ -152,14 +154,12 @@ const ActionsDAG::Node & addFunction(
 const ActionsDAG::Node & addCast(
         const ActionsDAGPtr & dag,
         const ActionsDAG::Node & node_to_cast,
-        const DataTypePtr & to_type,
-        OriginalToNewNodeMap & node_remap)
+        const DataTypePtr & to_type)
 {
     if (!node_to_cast.result_type->equals(*to_type))
```

Comment:
Should be `if (node_to_cast.result_type->equals(*to_type))` (without negation) to fix test `03271_sqllancer_having_issue.sql`? Didn't try, just a hunch.